src/FSharp.Control.TaskSeq.Test/TaskSeq.TakeWhile.Tests.fs

```diff
@@ -0,0 +1,135 @@
+module TaskSeq.Tests.TakeWhile
+
+open System
+open Xunit
+open FsUnit.Xunit
+open FsToolkit.ErrorHandling
+
+open FSharp.Control
+
+//
+// TaskSeq.takeWhile
+// TaskSeq.takeWhileAsync
+// TaskSeq.takeWhileInclusive
+// TaskSeq.takeWhileInclusiveAsync
+//
+
+module EmptySeq =
+    [<Theory; ClassData(typeof<TestEmptyVariants>)>]
+    let ``TaskSeq-takeWhile has no effect`` variant =
+        Gen.getEmptyVariant variant
+        |> TaskSeq.takeWhile ((=) 12)
+        |> TaskSeq.toListAsync
+        |> Task.map (List.isEmpty >> should be True)
+
+    [<Theory; ClassData(typeof<TestEmptyVariants>)>]
+    let ``TaskSeq-takeWhileAsync has no effect`` variant =
+        Gen.getEmptyVariant variant
+        |> TaskSeq.takeWhileAsync (fun x -> task { return x = 12 })
+        |> TaskSeq.toListAsync
+        |> Task.map (List.isEmpty >> should be True)
+
+// The primary requirement is that items after the item failing the predicate must be excluded
+module FiltersAfterFail =
+    [<Theory; InlineData false; InlineData true>]
+    let ``TaskSeq-takeWhile(Inclusive)? excludes all items after predicate fails`` inclusive =
+        // The only real difference in semantics between the base and the *Inclusive variant lies in whether the final item is returned
+        // NOTE the semantics are very clear on only propagating a single failing item in the inclusive case
+        let f, expected =
+            if inclusive then TaskSeq.takeWhileInclusive, "ABBC"
+            else TaskSeq.takeWhile, "ABB"
+        seq { 1; 2; 2; 3; 3; 2; 1 }
+        |> TaskSeq.ofSeq
+        |> f (fun x -> x <= 2)
+        |> TaskSeq.map char
+        |> TaskSeq.map ((+) '@')
+        |> TaskSeq.toArrayAsync
+        |> Task.map (String >> should equal expected)
+
+    // Same as preceding test, just with Async functions
+    [<Theory; InlineData false; InlineData true>]
+    let ``TaskSeq-takeWhile(Inclusive)?Async excludes all items after after predicate fails`` inclusive =
+        let f, expected =
+            if inclusive then TaskSeq.takeWhileInclusiveAsync, "ABBC"
+            else TaskSeq.takeWhileAsync, "ABB"
+        taskSeq { 1; 2; 2; 3; 3; 2; 1 }
+        |> f (fun x -> task { return x <= 2 })
+        |> TaskSeq.map char
+        |> TaskSeq.map ((+) '@')
+        |> TaskSeq.toArrayAsync
+        |> Task.map (String >> should equal expected)
+
+// Covers the fact that it's not sufficient to merely exclude successor items - it's also critical that the enumeration terminates
+module StopsEnumeratingAfterFail =
+    [<Theory; InlineData false; InlineData true>]
+    let ``TaskSeq-takeWhile(Inclusive)? stops consuming after predicate fails`` inclusive =
+        let f, expected =
+            if inclusive then TaskSeq.takeWhileInclusive, "ABBC"
+            else TaskSeq.takeWhile, "ABB"
+        seq { 1; 2; 2; 3; 3; failwith "Too far" }
+        |> TaskSeq.ofSeq
```

Comment:
I think I actually don't have a test like this currently, good to have it now. My approach is usually using mutables and doing plus-one between the items to ensure we stop at the proper point. But this works just fine as well.

Reply:
Thanks, I like your suggested way of doing it, it just hadn't occurred to me.